fdbserver/ptxn/TLogPeekCursor.actor.h

```diff
@@ -0,0 +1,314 @@
+
+#include "flow/actorcompiler.h" // has to be the last file included
+
+namespace ptxn {
+
+class PeekCursorBase {
+public:
+	// Iterator for pulled mutations. This is not a standard input iterator as it is not duplicable, thus no postfix
+	// operator++ support. NOTE It will *NOT* trigger remoteMoreAvailable(). It will only iterate over the serialized
+	// data that are already peeked from TLog.
+	// NOTE: Duplicating the iterator may led to unexpected behavior, i.e. step one iterator causes the other iterators
+	// being stepped. It is discouraged to explicitly duplicate the iterator.
+	class iterator : public ConstInputIteratorBase<VersionSubsequenceMutation> {
+		friend class PeekCursorBase;
+
+		PeekCursorBase* pCursor;
+		bool isEndIterator = false;
+
+		// pCursor_ is the pointer to the PeekCursorBase
+		// isEndIterator_ is used to determine if the iterator is served as the end of the stream. We never know where
+		// the end is when we iterate over the mutations, until we reached the end, thus we check if we are at the end
+		// of the stream when we compare the iterator with a special iterator pointing to nothing but having the flag
+		// isEndIterator on, and during the comparision we determine if the current iterator is reaching the end. See
+		// iterator::operator== for implementation details.
+		iterator(PeekCursorBase* pCursor_, bool isEndIterator_);
+
+	public:
+		bool operator==(const iterator&) const;
+		bool operator!=(const iterator&) const;
+
+		reference operator*() const;
+		pointer operator->() const;
+
+		// Prefix incremental
+		void operator++();
+
+		// Postfix incremental is disabled since duplication is prohibited.
+	};
+
+	PeekCursorBase();
+
+	// Checks if there is any more messages in the remote TLog(s), if so, retrieve the messages to local. Will
+	// invalidate the iterator, if exists.
+	Future<bool> remoteMoreAvailable();
+
+	// Gets one mutation
+	const VersionSubsequenceMutation& get() const;
+
+	// Moves to the next mutation, return false if there is no more mutation
+	void next();
+
+	// Any remaining mutation *LOCALLY* available
+	bool hasRemaining() const;
+
+	// Returns an iterator that represents the begin of the data still undeserialized.
+	iterator begin() { return iterator(this, false); }
+
+	// Returns an iterator that represents the end of the data
+	const iterator& end() const { return endIterator; }
+
+protected:
+	// Checks if there is any mutations remotely
+	virtual Future<bool> remoteMoreAvailableImpl() = 0;
+
+	// Steps the local cursor
+	virtual void nextImpl() = 0;
+
+	// Gets the message
+	virtual const VersionSubsequenceMutation& getImpl() const = 0;
+
+	// Checks if any remaining mutations
+	virtual bool hasRemainingImpl() const = 0;
+
+private:
+	// The iterator that represents the end of the unserialized data
+	const iterator endIterator;
+};
+
+// Connect to a given TLog server and peeks for mutations with a given TeamID
+class ServerTeamPeekCursor : public PeekCursorBase {
```

Comment:
We use TLog Groups to differentiate with Storage Server Teams.

```suggestion
class TLogGroupPeekCursor : public PeekCursorBase {
```